R/estimateDiversity.R

```diff
@@ -178,13 +188,108 @@ setMethod("estimateDiversity", signature = c(x = "SummarizedExperiment"),
                                          .get_diversity_values,
                                          x = x,
                                          mat = assay(x, abund_values),
+                                         tree = tree,
                                          BPPARAM = BPPARAM,
                                          ...)
         .add_values_to_colData(x, dvrsts, name)
     }
 )
 
+#' @rdname estimateDiversity
+#' @export
+setMethod("estimateDiversity", signature = c(x="TreeSummarizedExperiment"),
+    function(x, abund_values = "counts",
+             index = c("shannon","gini_simpson","inverse_simpson",
+                       "coverage", "fisher", "faith"),
+             name = index, tree = rowTree(x), ..., BPPARAM = SerialParam()){
+
+        # input check
+        # IF object does not have a tree
+        if( ("faith" %in% index) && (is.null(tree) || is.null(tree$edge.length)) ){
+            stop("Object does not have a tree or the tree does not have any branches.
+             'faith' is not possible to calculate.",
+                 call. = FALSE)
+        }
+
+        index<- match.arg(index, several.ok = TRUE)
+        if(!.is_non_empty_character(name) || length(name) != length(index)){
+            stop("'name' must be a non-empty character value and have the ",
+                 "same length than 'index'.",
+                 call. = FALSE)
+        }
+
+        if( "faith" %in% index ){
+            # Get the name of "faith" index
+            faith_name <- name[index %in% "faith"]
+            # And delete it from name
+            name <- name[!index %in% "faith"]
+
+            # Delete "faith" from indices
+            index <- index[!index %in% "faith"]
+        }
+
+        # If index list contained other than 'faith' index, the length of the list is over 0
+        if( length(index)>0){
+            # Calculates all indices but not 'faith'
+            x <- callNextMethod()
+        }
+
+        # If 'faith_name' exist, then 'faith' index was present in the index list
+        if( exists("faith_name") ){
```

Comment:
The `exists` function is a nice feature, but it is not the right solution in this instance. From the meaning of the check you want to answer the question "Do we need to calculate the faith index?". This can be answered with a "Yes" or "No" aka. boolean. Thus, it should be a scalar logical variable defining the answer to this question.

```r
calc_faith <- FALSE
if(...){
  ...
  calc_faith <- TRUE
}
...
if(calc_faith){
  ...
}
```

Also `exists` can walk environments, which leads to this behavior.

```
> FUN <- function(){
+     exists("abc")
+ }
> FUN()
[1] FALSE
> abc <- "meep"
> FUN()
[1] TRUE
```

If the user defines a variable `faith_name`, the index will always be tried to calculated. The `exists` function is therefore dangerous, especially if you don't define the `where` and/or `envir` argument. If would be very very careful with this function.

Reply:
Changed